warp10/src/main/java/io/warp10/script/ext/http/HTTP.java

```diff
@@ -0,0 +1,449 @@
+import io.warp10.warp.sdk.Capabilities;
+
+import org.apache.commons.codec.binary.Base64;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Send an HTTP request to an url
+ *
+ * To raise maximum number of calls and download size limit, use these capabilities:
+ * http.requests
+ * http.size
+ *
+ * Params:
+ * METHOD The http method
+ * URL The URL to send the request to. Must begin with http:// or https://
+ * HEADERS An optional header
+ * BODY An optional body. STRING or BYTES
+ * AUTH_INFO Authentication arguments. For example for basic authentication, provide [username, password]
+ * AUTH_MACRO A macro that expects " + AUTH_INFO + " on the stack, and returns a map to be appended with the headers. Default to basic authentication
+ * CHUNK_SIZE Chunk size
+ * CHUNK_MACRO A macro that is executed whenever a chunk has been downloaded. It expects a MAP that contains chunk number (a LONG), status code (a LONG), status message (a STRING), headers (a MAP), and chunk content (a BYTES objects)
+ *
+ * Output:
+ * RESPONSE A map that contains status code (a LONG), status message (a STRING), headers (a MAP) and full content of the response (a BYTES objects). The content is empty if chunk option is used
+ *
+ */
+public class HTTP extends NamedWarpScriptFunction implements WarpScriptStackFunction {
+
+  //
+  // Arguments
+  //
+
+  public static final String METHOD = "method";
+  public static final String URL = "url";
+  public static final String HEADERS = "headers";
+  public static final String BODY = "body";
+  public static final String AUTH_INFO = "auth.info";
+  public static final String AUTH_MACRO = "auth.macro";
+  public static final String CHUNK_SIZE = "chunk.size";
+  public static final String CHUNK_MACRO = "chunk.macro";
+
+  //
+  // Output
+  //
+
+  public static final String RESPONSE = "response";
+  public static final String STATUS_CODE = "status.code";
+  public static final String STATUS_MESSAGE = "status.message";
+  public static final String RESPONSE_HEADERS = "headers";
+  public static final String CONTENT = "content";
+  public static final String CHUNK_NUMBER = "chunk.number";
+
+  //
+  // Control
+  //
+
+  private final WebAccessController webAccessController;
+
+  //
+  // Authorization
+  //
+
+  private final boolean auth;
+  private final String capName;
+
+  //
+  // Limits
+  //
+
+  private final long baseMaxRequests;
+  private final long baseMaxSize;
+
+  public HTTP(String name) {
+    super(name);
+
+    String patternConf = WarpConfig.getProperty(HttpWarpScriptExtension.WARPSCRIPT_HTTP_HOST_PATTERNS);
```

Comment:
Parameter extraction should be done in a static way so other instances of HTTP do not repeat it.